src/pfs/pfs.go

```diff
@@ -343,51 +347,91 @@ func onlyHex(p []byte) bool {
 	return true
 }
 
+type ancestryRegexp struct {
+	*regexp.Regexp
+	partMap map[string][]byte
+}
+
+func makeAncestryRegexp(ancestryPart []byte) (ancestryRegexp, error) {
+	// ?P<> syntax assigns each capture group a name.
+	re := regexp.MustCompile(`^(?P<dot>\.\d+)?((?P<carets>\^+)|(?P<caretWithNumbers>\^\d+)|(?P<tildes>~+)|(?P<tildeWithNumbers>~\d+))?$`)
+	if !re.Match(ancestryPart) {
+		return ancestryRegexp{}, errors.New("invalid ancestry format")
+	}
+	parts := re.FindSubmatch(ancestryPart)
+	partMap := make(map[string][]byte) // a partMap is easier to work with.
+	partNames := re.SubexpNames()
+	for i, part := range parts {
+		partMap[partNames[i]] = part
+	}
+	ancestryRe := ancestryRegexp{
+		Regexp:  re,
+		partMap: partMap,
+	}
+	return ancestryRe, nil
+}
+
+const (
+	dot              = "dot"
+	carets           = "carets"
+	caretWithNumbers = "caretWithNumbers"
+	tildes           = "tildes"
+	tildeWithNumbers = "tildeWithNumbers"
+)
+
 // counts the overall offset of branch root and ancestor of
 // also removes everything of the ancestry reference
 // this function also check following circumstances for ancestry reference:
-// 1. ancestry reference should only contain "^" "." and digits
+// 1. ancestry reference should only contain "^", "~", ".", and digits
```

Comment:
We intentionally omitted some of these during our initial implementation so we could get a subset of parsing working for human use cases.